stdlib/src/collections/deque.mojo

```diff
@@ -0,0 +1,425 @@
+from documentation import doc_private
+from memory import UnsafePointer
+
+
+# ===----------------------------------------------------------------------===#
+# Deque
+# ===----------------------------------------------------------------------===#
+
+
+struct Deque[ElementType: CollectionElement](
+    Movable, ExplicitlyCopyable, Sized, Boolable
+):
+    """Implements a double-ended queue.
+
+    It supports pushing and popping from both ends in O(1) time resizing the
+    underlying storage as needed.
+
+    Parameters:
+        ElementType: The type of the elements in the deque.
+            Must implement the trait `CollectionElement`.
+    """
+
+    # ===-------------------------------------------------------------------===#
+    # Aliases
+    # ===-------------------------------------------------------------------===#
+
+    alias default_capacity: Int = 64
+    """The default capacity of the deque: must be the power of 2."""
+
+    # ===-------------------------------------------------------------------===#
+    # Fields
+    # ===-------------------------------------------------------------------===#
+
+    var _data: UnsafePointer[ElementType]
+    """The underlying storage for the deque."""
+
+    var _head: Int
+    """The index of the head: points the first element of the deque."""
+
+    var _tail: Int
+    """The index of the tail: points behind the last element of the deque."""
+
+    var _capacity: Int
+    """The amount of elements that can fit in the deque without resizing it."""
+
+    var _min_capacity: Int
+    """The minimum required capacity in the number of elements of the deque."""
+
+    var _maxlen: Int
+    """The maximum number of elements allowed in the deque.
+
+    If more elements are pushed, causing the total to exceed this limit,
+    items will be popped from the opposite end to maintain the maximum length.
+    """
+
+    var _shrink: Bool
+    """ Indicates whether the deque's storage is re-allocated to a smaller size when possible."""
+
+    # ===-------------------------------------------------------------------===#
+    # Life cycle methods
+    # ===-------------------------------------------------------------------===#
+
+    fn __init__(
+        inout self,
+        *,
+        owned elements: Optional[List[ElementType]] = None,
+        capacity: Int = self.default_capacity,
+        min_capacity: Int = self.default_capacity,
+        maxlen: Int = -1,
+        shrink: Bool = True,
+    ):
+        """Constructs a deque.
+
+        Args:
+            elements: The optional list of initial deque elements.
+            capacity: The initial capacity of the deque.
+            min_capacity: The minimum allowed capacity of the deque when shrinking.
+            maxlen: The maximum allowed capacity of the deque when growing.
+            shrink: Should storage be de-allocated when not needed.
+        """
+        if capacity <= 0:
+            deque_capacity = self.default_capacity
```

Comment:
**Suggestion** Here and elsewhere: while we allow the lack-of-use of `var`, internally within the standard library, we'd like to keep using `var`.  We discussed this a few weeks ago and will solidify this in our documented style guide to avoid confusion.  Do you mind changing this in a follow-up PR?  No need to block this PR on it.

(FYI @martinvuyk as I know you were asking about this in a different PR of yours).

Reply:
Ok, will add `var` in the next PR where it is expected.